server/base/src/main/java/org/apache/accumulo/server/metadata/RootGcCandidates.java

```diff
@@ -0,0 +1,120 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.accumulo.server.metadata;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.util.Collection;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Stream;
+
+import org.apache.accumulo.core.metadata.schema.Ample.FileMeta;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+public class RootGcCandidates {
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+
+  // This class is used to serialize and deserialize root tablet metadata using GSon. Any changes to
+  // this class must consider persisted data.
+  private static class GSonData {
+    int version = 1;
+
+    // SortedMap<dir path, SortedSet<file name>>
+    SortedMap<String,SortedSet<String>> candidates;
+  }
+
+  /*
+   * The root tablet will only have a single dir on each volume. Therefore root file paths will have
+   * a small set of unique prefixes. The following map is structured to avoid storing the same dir
+   * prefix over and over in JSon and java.
+   *
+   * SortedMap<dir path, SortedSet<file name>>
+   */
+  private SortedMap<String,SortedSet<String>> candidates;
+
+  public RootGcCandidates() {
+    this.candidates = new TreeMap<>();
+  }
+
+  private RootGcCandidates(SortedMap<String,SortedSet<String>> candidates) {
+    this.candidates = candidates;
+  }
+
+  public void add(Collection<? extends FileMeta> refs) {
+    refs.forEach(ref -> {
+      Path path = ref.path();
+
+      String parent = path.getParent().toString();
+      String name = path.getName();
+
+      candidates.computeIfAbsent(parent, k -> new TreeSet<>()).add(name);
+    });
+  }
+
+  public void remove(Collection<String> refs) {
+    refs.forEach(ref -> {
+      Path path = new Path(ref);
+      String parent = path.getParent().toString();
+      String name = path.getName();
+
+      SortedSet<String> names = candidates.get(parent);
+      if (names != null) {
+        names.remove(name);
+        if (names.isEmpty()) {
+          candidates.remove(parent);
+        }
+      }
+    });
+  }
+
+  public Stream<String> stream() {
+    return candidates.entrySet().stream().flatMap(entry -> {
+      String parent = entry.getKey();
+      SortedSet<String> names = entry.getValue();
+      return names.stream().map(name -> new Path(parent, name).toString());
+    });
+  }
+
+  public String toJson() {
+    GSonData gd = new GSonData();
+    gd.candidates = candidates;
+
+    // replace two space indentation with one space to save bytes in ZK while still leaving it human
+    // readable
+    return GSON.toJson(gd).replaceAll("  ", " ");
```

Comment:
I think tweaking the pretty-print is error prone... this could affect values. We should either store minimized (my preference; can use jq or pretty print on display), or store pretty-print without modifications. Gson may also have an option to change the indent depth for its pretty-printer... which gets you the best of both.

Could also just make this less error prone by replacing only spaces at the beginning of the line... but I don't know if that might still have issues on wrapped strings in JSON.

Reply:
That could be error prone, I will remove it.

Reply:
Removed it.